ssl/quic/quic_wire_pkt.c

```diff
@@ -0,0 +1,734 @@
+            || !PACKET_copy_bytes(pkt, hdr->src_conn_id.id, src_conn_id_len))
+            return 0;
+
+        hdr->version            = (uint32_t)version;
+        hdr->dst_conn_id.id_len = (unsigned char)dst_conn_id_len;
+        hdr->src_conn_id.id_len = (unsigned char)src_conn_id_len;
+
+        if (version == 0) {
+            /*
+             * Version negotiation packet. Version negotiation packets are
+             * identified by a version field of 0 and the type bits in the first
+             * byte are ignored (they may take any value, and we ignore them).
+             */
+            hdr->type       = QUIC_PKT_TYPE_VERSION_NEG;
+            hdr->fixed      = (b0 & 0x40) != 0;
+
+            hdr->data       = PACKET_data(pkt);
+            hdr->len        = PACKET_remaining(pkt);
+
+            /* Version negotiation packets are always fully decoded. */
+            hdr->partial    = 0;
+
+            /* Fields not used in version negotiation packets. */
+            hdr->pn_len             = 0;
+            hdr->spin_bit           = 0;
+            hdr->key_phase          = 0;
+            hdr->token              = NULL;
+            hdr->token_len          = 0;
+            memset(hdr->pn, 0, sizeof(hdr->pn));
+
+            if (!PACKET_forward(pkt, hdr->len))
+                return 0;
+        } else if (version != QUIC_VERSION_1) {
+            /* Unknown version, do not decode. */
+            return 0;
+        } else {
+            if (l < QUIC_MIN_VALID_PKT_LEN_CRYPTO)
+                return 0;
+
+            /* Get long packet type and decode to QUIC_PKT_TYPE_*. */
+            raw_type = ((b0 >> 4) & 0x3);
+
+            switch (raw_type) {
+                case 0: hdr->type = QUIC_PKT_TYPE_INITIAL; break;
+                case 1: hdr->type = QUIC_PKT_TYPE_0RTT; break;
+                case 2: hdr->type = QUIC_PKT_TYPE_HANDSHAKE; break;
+                case 3: hdr->type = QUIC_PKT_TYPE_RETRY; break;
+            }
+
+            hdr->pn_len     = 0;
+            hdr->fixed      = 1;
+
+            /* Fields not used in long-header packets. */
+            hdr->spin_bit   = 0;
+            hdr->key_phase  = 0;
+
+            if (hdr->type == QUIC_PKT_TYPE_INITIAL) {
+                /* Initial packet. */
+                uint64_t token_len;
+
+                if (!PACKET_get_quic_vlint(pkt, &token_len)
+                    || token_len > SIZE_MAX
+                    || !PACKET_get_bytes(pkt, &hdr->token, token_len))
+                    return 0;
+
+                hdr->token_len  = (size_t)token_len;
+                if (token_len == 0)
+                    hdr->token = NULL;
+            } else {
+                hdr->token      = NULL;
+                hdr->token_len  = 0;
+            }
+
+            if (hdr->type == QUIC_PKT_TYPE_RETRY) {
+                /* Retry packet. */
+                hdr->data       = PACKET_data(pkt);
+                hdr->len        = PACKET_remaining(pkt);
```

Comment:
Uhmmm, unless I'm reading https://datatracker.ietf.org/doc/html/rfc9000#section-17.2.5 wrong, the RETRY packet has no payload.  It does, however, have a token (which should end up in `hdr->token`) and a token integrity tag.

Reply:
Yes, the RFC calls it a retry token plus retry integrity instead of a payload. In actual functional terms this is the payload of the packet, so my packet decoder takes the pragmatic view of just considering this to be the payload of a Retry packet. I don't think it makes sense to have separate fields for this in the header structure.

Reply:
Point is that you already have the `token` field...  that's where I'd expect to find the retry token!

Reply:
That's found in the Initial packet header and is something very different... the Retry Token resembles a payload much more than it resembles the Initial Token.

Reply:
Ok, fair enough